node/src/env0-deploy-utils.js

```diff
@@ -50,49 +50,105 @@ class DeployUtils {
     }
 
     console.log(`setting the following configuration: ${JSON.stringify(configuration)}`);
-    await apiClient.callApi('post', 'configuration', {data: {...configuration, projectId: undefined}});
+    await this.apiClient.callApi('post', 'configuration', {data: {...configuration, projectId: undefined}});
   }
 
   async deployEnvironment(environment, blueprintRevision, blueprintId) {
-    console.log(`starting to deploy environmentId: ${environment.id}, blueprintId: ${blueprintId}`);
-    const deployment = await apiClient.callApi('post', `environments/${environment.id}/deployments`,
-        {data: {blueprintId, blueprintRevision}});
+    await this.waitForEnvironment(environment.id);
 
-    console.log(`Started deployment ${deployment.id}`);
+    return await this.apiClient.callApi('post', `environments/${environment.id}/deployments`, {data: {blueprintId, blueprintRevision}});
   }
 
   async destroyEnvironment(environment) {
-    console.log(`Starting to destroy environmentId: ${environment.id}`);
-    const deployment = await apiClient.callApi('post', `environments/${environment.id}/destroy`);
-    console.log(`Started destroy ${deployment.id}`);
+    await this.waitForEnvironment(environment.id);
+
+    return await this.apiClient.callApi('post', `environments/${environment.id}/destroy`);
   }
 
+  async writeDeploymentStepLog(deploymentLogId, stepName) {
+    let shouldPoll, startTime;
 
-  async pollEnvironmentStatus(environmentId) {
-    const deployEndStatuses = ['CREATED', 'ACTIVE', 'INACTIVE', 'FAILED', 'TIMEOUT'];
-    const maxRetryNumber = 90; //waiting for 15 minutes (90 * 10 seconds)
-    let retryCount = 0;
+    do {
+      const steps = await this.apiClient.callApi('get', `deployments/${deploymentLogId}/steps`);
+      const step = steps.find(step => step.name === stepName);
 
-    console.log(`Starting status polling for environment status for ${environmentId}, maxRetryNumber: ${maxRetryNumber}`);
-    // eslint-disable-next-line no-constant-condition
-    while (true) {
-      const environment = await apiClient.callApi('get', `environments/${environmentId}`);
-      if (deployEndStatuses.includes(environment.status)) {
-        return environment.status;
-      } else if (retryCount >= maxRetryNumber) {
-        throw new Error('Polling environment reached max retries');
+      const { events, nextStartTime, hasMoreLogs } = await this.apiClient.callApi(
+          'get',
+          `deployments/${deploymentLogId}/steps/${stepName}/log`,
+          { params: { startTime }}
+          );
+
+      events.forEach((event) => console.log(event.message));
+
+      if (nextStartTime) startTime = nextStartTime;
+
+      shouldPoll = hasMoreLogs || step.status === 'IN_PROGRESS';
+    } while (shouldPoll)
```

Comment:
If I'm not mistaken polling without timeout on `hasMoreLogs` is OK, because there are already more logs available. However polling again when there are no more logs and the status is `IN_PROGRESS` should have some sleep in here.

Take a look at `getDeploymentStepLog` in `deployment-steps.store.ts` for reference

Reply:
b29e73e